### ISSUE TITLE
naiveSyncBN on Waymo causes issues with samples_per_gpu = 2

### ISSUE THREAD

**Issue by Divadi** [open-mmlab/mmdetection3d#884]
**Describe the bug**
When I run the default waymo config for SECOND
https://github.com/open-mmlab/mmdetection3d/blob/master/configs/second/hv_second_secfpn_sbn_2x16_2x_waymoD5-3d-3class.py
with a batch size per gpu of 2, my linux server completely crashes. I have tried changing a variety of hyper-parameters, but only changing naive sync BN to regular BN seems to resolve this issue.
What could be the cause of this?

**Environment**
```
sys.platform: linux
Python: 3.6.13 |Anaconda, Inc.| (default, Jun  4 2021, 14:25:59) [GCC 7.5.0]
CUDA available: True
GPU 0,1,2,3: RTX A6000
CUDA_HOME: /usr
NVCC: Build cuda_11.1.TC455_06.29190527_0
GCC: gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
PyTorch: 1.8.0
PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2019.0.4 Product Build 20190411 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v1.7.0 (Git Hash 7aed236906b1f7a05c0917e5257a1af05e9ff683)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.1
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=compute_37
  - CuDNN 8.0.5
  - Magma 2.5.2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.1, CUDNN_VERSION=8.0.5, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.8.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON,

TorchVision: 0.9.0
OpenCV: 4.5.3
MMCV: 1.3.9
MMCV Compiler: GCC 7.3
MMCV CUDA Compiler: 11.1
MMDetection: 2.14.0
MMSegmentation: 0.14.1
MMDetection3D: 0.15.0+7eb9b8c
```
**Error traceback**
I have checked syslogs and dmesg but nothing relevant could be found there.
In the past, I have had issues (albeit on a different machine, different model, different dataset) of SyncBN causing training to just hang.

**Comment by Tai-Wang** [comment]
That's really strange. We've never met this before. I guess the reason should be subtle and possibly relevant to your setup. If there is no valuable information in the traceback, we can only look forward to voices from the community.

**Comment by Divadi** [comment]
One thing to note is that I had --deterministic on - I'm not sure if that would affect things.
I will try it sometime without that flag

**Comment by zhanggefan** [comment]
Hi, @Divadi 
Have you tried replacing NaiveSyncBN with SyncBN? For torch1.8 the native syncbn is more stable and memory-efficient than NaiveSyncBN. You may have a try.

**Comment by Divadi** [comment]
@zhanggefan Ah, so it's fixed now? I remember reading that the default torch syncbn gave poor results.
I will give it a try

**Comment by zhanggefan** [comment]
@Divadi 

If I remember right, the critical performance issue concerning tracking states updates is fixed by [PR#22248](https://github.com/pytorch/pytorch/pull/22248) in PyTorch v1.2, though I am not sure if there exist other unknown bugs. 

But as I have done experiments of point-pillars on Waymo dataset several times (with both torch v1.6 & v1.8, fp16 enabled), I can guarantee you that (at least for point pillars which shares the same backbone as SECOND) replacing NaiveSyncBN with SyncBN causes no performance drop.

Let's go back to the crashing issue. After days of thinking now, I also don't think that NaiveSyncBN is the root cause of the crash. If nothing special is left in syslog after the crash, then it is more likely a hardware problem. I encountered a similar issue these days when I trained yolov5 on an image object detection task. My server always rebooted within a hundred iterations. After some thorough investigation (by swapping hardware), we found that it is the SuperMicro's barebone (including only motherboard X11DPG-QT and 2 * PSU) that caused the problem. When we swapped to another barebone from another production batch, the reboot was gone completely.

Hope this information helps.

**Comment by Divadi** [comment]
Thank you for the advice
Unfortunately, I really have no way of deconstructing the hardware. I have also tried the native syncbn, but that crashed too (after lasting a little longer).
I'll close the issue for now as it might be a hardware issue.
